Add background color control for table block

talldan opened the pull request:
## Description
Adds background and text color controls to the table block.

Works with both Regular and Stripes style variations of the table block.

## How has this been tested?
- Manual testing

## Screenshots <!-- if applicable -->
![screen shot 2018-12-28 at 5 45 31 pm](https://user-images.githubusercontent.com/677833/50511673-8950f800-0ac9-11e9-8055-0b1455b60e56.png)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

talldan committed:
Add background and text color controls for table block

talldan committed:
Modify fallback styles to use first table row

talldan committed:
Make text color control apply to all rows, not only striped row

talldan commented on packages/block-library/src/table/index.js:
```diff
@@ -111,17 +129,31 @@ export const settings = {
 			'has-fixed-layout': hasFixedLayout,
 		} );
 
+		const isStriped = !! className && className.indexOf( 'is-style-stripes' ) > -1;
```
Not a fan of having to detect the style variation like this, I wonder if there's a better way?

ZebulanStanphill replied:
@talldan Use `className.includes( 'is-style-stripes' )`.

ZebulanStanphill replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

talldan replied:
@ZebulanStanphill Cheers for the suggestion, though I was referring to checking for values in strings being a brittle implementation rather than the method used to check them - see http://wiki.c2.com/?StringlyTyped.

ZebulanStanphill replied:
Ah, I see.

Perhaps there should be a `getStyleClass` function defined somewhere that you could pass the style name into, which would return the class for that style, abstracting away the `is-style-` part.

Or perhaps there should be a `getBlockStyle` function that determines if a block is using a style variation somehow. Of course, the only way to check right now is via classes, but if it was later decided to include style variation names in the comment delimiters or something like that, only that function would have to be changed.

talldan replied:
I've changed the implementation now to use a different approach and this code no longer exists, so lets not worry about it for now 😄

talldan commented on packages/block-library/src/table/edit.js:
```diff
@@ -407,13 +458,45 @@ export default class TableEdit extends Component {
+							},
+						] }
+					>
+						<ContrastChecker
```
Contrast checker doesn't work so well with the stripes variation, since on some rows the contrast might be fine, but on others the contract could be problematic. One way to solve could be to pass through multiple colors in an array? Maybe something like an `additionalColors` prop. Alternatively, use two contrast checkers? That could get confusing.

talldan commented on packages/block-library/src/table/edit.js:
```diff
@@ -30,7 +39,18 @@ import {
+
+const applyFallbackStyles = withFallbackStyles( ( node, props ) => {
+	const { textColor, backgroundColor } = props.attributes;
+	const styles = node ? getComputedStyle( node.querySelector( 'tr' ) ) : undefined;
```
Currently only grabs fallback colors for the first row.

karmatosed reviewed (dismissed):
Let's scale this back a little to just having the background color and not a text color also. We should only have this for the stripes not for every cell right now. I have some ideas how we can do this and will comment in issue but let's also make sure this gets a 'design feedback' label as this goes in for Wednesday because a UI change.

karmatosed commented:
We need to remove text color options and focus just on background ones that can't become inaccessible, just for stripes. As part of this, we should remove the color picker.

Color palette wise let's have it use a fixed palette. We have a few options:

1. Use a light palette based on a set of colors:
![2018-10-15 at 12 39](https://user-images.githubusercontent.com/253067/46949076-0c8cb100-d078-11e8-92d1-e3fa7a0b1f04.png)

2. Desaturate and lighten the existing theme palette:

![2018-10-15 at 12 47](https://user-images.githubusercontent.com/253067/46949219-950b5180-d078-11e8-909d-ac1fcb8d7c67.png)

Ideally, we would go for 2 but with time stop of Wednesday, let's make sure we get something in.

ZebulanStanphill commented:
What if the stripes were a lighter or darker variant of the selected background color? Would that idea work?

postphotos commented:
> What if the stripes were a lighter or darker variant of the selected background color? Would that idea work?

That could work as well!

talldan committed:
Simplify implementation of colors for table block.

- Remove text color control
- Disable custom colors
- Apply class names to table element instead of tr element
- Remove JS striping

talldan committed:
Filter colors in table background color control to only those that contrast with the default text color

talldan commented:
I've refactored this quite a bit now from yesterday:

- Removed text color controls.
- Removed the ability to choose custom background colors.
- Dynamically filtered out background colors that don't contrast with the default text color. (potentially though, this might exclude all colors for some themes? We could hide the color control for those themes.)
- Use CSS only for styling background colors.

I think this makes the implementation a lot cleaner, and hopefully mergeable, but would love @jorgefilipecosta's feedback on the changes to color utils, ColorPalette and withColorContext.

talldan commented:
Hi @karmatosed - My understanding is that the colors can be provided by the theme, so potentially the second option would be preferable so that the colors are still in keeping with the theme. Having said that, I'm not sure how we can lighten the colors dynamically if they're specified through css.

I think we'd have to only set the colors using inline styles (instead of classes) and lighten the colors in JavaScript. Potentially that could work, but might need input from a dev with more experience of themes / the color system.

At the moment I've used a different approach - I've filtered the color palette down to only accessible colors, which means the red and black colors are no longer available:
![screen shot 2018-10-16 at 1 25 27 pm](https://user-images.githubusercontent.com/677833/46994451-15988380-d147-11e8-8184-bc1175578392.png)

I know this is not a particularly great set of colors to choose from. Three or four of them are ok.

jorgefilipecosta commented on packages/editor/src/components/colors/utils.js:
```diff
@@ -72,3 +72,19 @@ export function getMostReadableColor( colors, colorValue ) {
+*
+* @return {Array} Array of filtered colors
+*/
+export function getContrastingColors( colors, contrastingColor ) {
```
We already have a getMostReadableColor, maybe we should call this function getReadableColors.

jorgefilipecosta commented on packages/editor/src/components/color-palette/with-color-context.js:
```diff
@@ -12,10 +12,10 @@ import { withSelect } from '@wordpress/data';
 			const settings = select( 'core/editor' ).getEditorSettings();
 			const colors = settings.colors;
-			const disableCustomColors = settings.disableCustomColors;
+			const disableCustomColors = settings.disableCustomColors || ownProps.disableCustomColors;
```
In PR https://github.com/WordPress/gutenberg/pull/10457 I'm adding the possibility to overwrite color definitions for a specific component. If we manage the land it before it this PR it may be very useful and simplify the changes here.

jorgefilipecosta commented on packages/components/src/color-palette/index.js:
```diff
@@ -17,15 +17,17 @@ import Button from '../button';
 import Tooltip from '../tooltip';
 
-export default function ColorPalette( { colors, disableCustomColors = false, value, onChange, className } ) {
+export default function ColorPalette( { colors, disableCustomColors = false, value, onChange, className, filterColors } ) {
```
To me filtering colors does not seem a responsibility of the ColorPalette component, why not pass the colors already filtered?

jorgefilipecosta reviewed (commented):
Hi @talldan I think if we manage to merge https://github.com/WordPress/gutenberg/pull/10457 before this PR we may be able to simplify the changes here as it looks like the other PR is addressing a need from this PR.
I really like the feature being added here 👍
In the last version when I try to execute this branch I'm getting an error:
```
TypeError: Failed to execute 'getComputedStyle' on 'Window': parameter 1 is not of type 'Element'.
    at edit.js:47
```

ZebulanStanphill commented on packages/block-library/src/table/edit.js:
```diff
@@ -417,3 +469,10 @@ export default class TableEdit extends Component {
 	}
 }
+
+const ComposedTableEdit = compose( [
```
In other block files I've seen the composed version be defined on the same line it is exported (so it is not given a name). Just noting for the sake of consistency.

ZebulanStanphill commented on packages/block-library/src/table/edit.js:
```diff
@@ -325,12 +346,13 @@ export default class TableEdit extends Component {
-							const classes = classnames( {
-								'is-selected': isSelected,
-							} );
+							const cellClasses = classnames( { 'is-selected': isSelected	} );
```
It looks like there is a stray tab character here.

ZebulanStanphill commented on packages/block-library/src/table/edit.js:
```diff
@@ -7,7 +7,15 @@ import classnames from 'classnames';
  * WordPress dependencies
  */
 import { Fragment, Component } from '@wordpress/element';
```
Nitpick: it would be nice if imports were ordered alphabetically. They tend to be in a random order that varies from file to file currently.

talldan replied:
Seems a bit unnecessary when there are just two imports.

talldan committed:
Remove errant tab

talldan committed:
Export using a single statement

talldan committed:
Revert "Filter colors in table background color control to only those that contrast with the default text color"

This reverts commit 8a39fbcf4ce4e6066a16c744c6e0a01936bc542f.

talldan committed:
Fix issue with row element not being present when fallback styles first applied

talldan committed:
Use a predefined set of background colors suitable for a table

jorgefilipecosta reviewed (commented):
Hi @talldan this PR is testing well for me, nice work 👍
I think it would be perfect if we could avoid the use of inline styles I created a PR and left a comment with a possible path to get us there.

talldan commented on packages/block-library/src/table/edit.js:
```diff
@@ -30,7 +39,54 @@ import {
-export default class TableEdit extends Component {
+const { getComputedStyle } = window;
+
+const BACKGROUND_COLORS = [
```
Using classes would be great. 

I do wonder about backwards compatibility in the case where we decide to update the custom colors to something else. I suppose a block with an old color would continue to work as long as we retain the class names defining the colors?

jorgefilipecosta committed:
Add createCustomColorsHOC HOC builder that creates withColors HOC for usages with custom colors array.

Refactor withColorHOC. Ensure withSelect is not used when not needed, and ensure custom color HOC has the right display name

Add tests for new createCustomColorsHOC

karmatosed commented:
@talldan could we distill back to just the 4 color options then?

![46949076-0c8cb100-d078-11e8-92d1-e3fa7a0b1f04](https://user-images.githubusercontent.com/253067/47618154-951f3e80-dac7-11e8-8561-ff74a92545fa.png)

talldan committed:
Use createCustomColorHOC for table backgrounds and refactor to use only class names to set the background color

talldan commented:
I've brought this back up to date, switched to the four colors defined by @karmatosed, and have also refactored to using class names instead of inline styles.

Currently I've cherry-picked the code from https://github.com/WordPress/gutenberg/pull/10945 into this branch, so we'd need to get that reviewed and merged first, then I'll rebase this.

![screen shot 2018-12-28 at 5 45 31 pm](https://user-images.githubusercontent.com/677833/50511641-60306780-0ac9-11e9-9c9c-ab1cbe868ab1.png)

talldan committed:
Remove contrast checker from table block background color selector - table block has custom background colors designed to meet contrast requirements

noisysocks commented:
Removing the `Needs Design Feedback` label as it looks like @karmatosed has given her feedback here.

noisysocks reviewed (approved):
LGTM, minor comments aside 👍 

I was confused by the `createCustomColorsHOC` API at first but the JSDoc code examples helped me eventually grok it. I think that using render props here might be simpler than a HOC but then again I think that about most of our HOCs, so... :shipit:😀

talldan committed:
Remove use of compose for single HOC

talldan committed:
Modify use of classnames, pass backgroundColor as an argument instead of using the object form

talldan commented on packages/block-library/src/table/edit.js:
```diff
@@ -421,3 +474,7 @@ export default class TableEdit extends Component {
+
+export default compose( [
+	withCustomBackgroundColors( 'backgroundColor' ),
+] )( TableEdit );
```
Very true, addressed in 38a74ad

talldan commented on packages/block-library/src/table/index.js:
```diff
@@ -113,15 +116,25 @@ export const settings = {
 			'has-fixed-layout': hasFixedLayout,
+			'has-background': !! backgroundClass,
+			[ backgroundClass ]: backgroundClass,
 		} );
```
Good spot! Addressed in 8fddd40

talldan committed:
update CHANGELOG

youknowriad commented:
👍 Nice
